src/audio/kpb.c

```diff
@@ -1000,19 +1160,38 @@ static int kpb_copy(struct comp_dev *dev)
 			break;
 		}
 
-		kpb_copy_samples(sink_c, source_c, copy_bytes, sample_width, channels);
-
+		if (kpb->num_of_sel_mic == 0) {
+			kpb_copy_samples(sink_c, source_c, copy_bytes, sample_width, channels);
+		} else {
+			uint32_t avail = audio_stream_get_avail_bytes(&source_c->stream);
+			uint32_t free = audio_stream_get_free_bytes(&sink_c->stream);
+
+			copy_bytes = MIN(avail, free * channels / kpb->num_of_sel_mic);
+			copy_bytes = ROUND_DOWN(copy_bytes, (sample_width >> 3) * channels);
+			produced_bytes = copy_bytes * kpb->num_of_sel_mic / channels;
+			produced_bytes =
+				ROUND_DOWN(produced_bytes,
+					   (sample_width >> 3) * kpb->num_of_sel_mic);
```

Comment:
I'm wondering if the compiler is smart enough to calculate this only once, maybe we could help it a bit by doing `unsigned int total_bytes_per_sample = (sample_width >> 3) * kpb->num_of_sel_mic;`